Add clock driver for meson

terryzbai opened the pull request:
This clock driver configures all clocks as needed at boot time. The initial configurations are parsed from device tree with a python script (which will replaced with either Zig or Rust) and saved as a C header file to be accessed by clock driver.

The design of clock driver is derived from Linux Common Clock Framework (CCF) that unifies client interfaces and key data structures but simplified for modularity and maintainability. Additionally, clock definitions and operations are implemented as static to avoid dynamic memory allocation and other dangerous behaviours.

The current implementation assumes that all client requests are safe and compatible with the existing configurations, especially for `clk_set_rate()`, which simply adjust clock rates or propagate one level up to parent clock without any checking. A virtualiser is probably needed to manage requests according to the safety policy. `determine_rate()` and `round_rate()` need to be implemented to consider more complicated cases, such as multiple clock consumers requesting different frequencies, and dynamically clock scaling for power management.

Linux clock driver is the main reference but it is not always compatible with what described in the datasheet, and it could go through some obscure code somewhere, so there might be some unknown errors in my code.

~~The clock driver for maaxboard is in progress and coming soon.~~
The implementation for maaxboard is also ready.

terryzbai committed:
Clock driver for odroidc4

This implementation includes all clock definitions, basic operations
of clock compoents, as well as interfaces to interact with clicnets
via PPC. Client should be able to enable/disable a specific gate
clock or query the current clock rate by passing the corresponding
clock identifier, which is listed in g12a-bindings.h.

set_rate() assumes that the rate requst is valid and compatible with
the current configurations. Complete implementation that considers
more incompatible requests will come in the future when dynamic
clock configuration is necessary in lionsos

Signed-off-by: Terry Bai <terry.z.bai@gmail.com>

terryzbai committed:
add Makefile

Signed-off-by: Terry Bai <terry.z.bai@gmail.com>

terryzbai committed:
Clock driver for maaxboard

Add clock driver for maaxboard. This commit includes all clock component
definitions, basic operations, and a simple test on maaxboard. Scripts are
added in build systems. A few files are reorganised for reuse

Signed-off-by: Terry Bai <terry.z.bai@gmail.com>

terryzbai committed:
Clock driver for maaxboard

Add clock driver for maaxboard. This commit includes all clock component
definitions, basic operations, and a simple test on maaxboard. Scripts are
added in build systems. A few files are reorganised for reuse

Signed-off-by: Terry Bai <terry.z.bai@gmail.com>

terryzbai committed:
Clk: fix style

Signed-off-by: Terry Bai <terry.z.bai@gmail.com>

terryzbai committed:
clk: modify interfaces and fix issues during initialisation

This commit includes modifiation to the interfaces, primarily for set/get rate, which
return error state as well as final rate.

More tests are conducted to verify the basic operations of types of clock components.
A few bugs are fixed as well.

Signed-off-by: Terry Bai <terry.z.bai@gmail.com>

terryzbai committed:
clk: fix style and add licenses

This commit fixed style and license issues

Signed-off-by: Terry Bai <terry.z.bai@gmail.com>

wom-bat reviewed (changes requested):
These are too close to the Linux drivers.  We don't need to provide a run-time interface for different clocks, as any sDDF system runs only on a single SoC --- so get rid of all the `..._ops` structures and call the functions directly.  

Plus please make sure all lines are under 80 chars wide, and add copyright and SPDX Licence tags where they;re missing.  Also remember to sign-off your commit messages.

terryzbai commented on drivers/clk/imx/include/clk-imx.h:
```diff
@@ -0,0 +1,332 @@
+ * structures at compile time, and will be replaced with
+ * real base addresses in clk_probe().
+ */
+#define CCM_BASE 0x1234
```
> We shouldn't need this, as we know at build time what the addresses are.

There are two clock-related memory regions on imx8m* boards: CCM (0x3038_0000) and CCM_ANALOG (0x3036_0000). A clock control register could be within either one of both. 

In my implementation, all clock structures are defined as static structures so these two magic numbers are set at compile time as specifiers. Once the driver knows the real address at build time,  it replaces the magic numbers in `clk_probe()`.

In my test, without these two magic numbers, all base addresses are set to NULL (0x0).

terryzbai commented on drivers/clk/clk.h:
```diff
@@ -0,0 +1,399 @@
+ *
+ */
+struct clk_ops {
+    uint8_t (*get_parent)(const struct clk *clk);
```
> If these operations do not vary per clock, get rid of the vector.

These operations vary per type of clock. We need this structure to redirect operation call to the clock-specific implementation instead of writing a bunch of conditional statements...

terryzbai commented on drivers/clk/clk.h:
```diff
@@ -0,0 +1,399 @@
+ *    of this register, and mask of gate bits are in higher 16-bit of this
+ *    register.  While setting the gate bits, higher 16-bit should also be
+ *    updated to indicate changing gate bits.
+ * CLK_GATE_BIG_ENDIAN - by default little endian register accesses are used for
```
> Do we have any of these big-endian registers for our hardware? Get rid of if not.

We don't, but I could not tell what hardware features we would have, so these definitions were just copied from Linux for scalability. Will check what could be removed for simplicity.

terryzbai committed:
clk: fix style and data types

This commit replaces all platform-dependent data types with fix-sized
types, and unifies the return values of interfaces as signed integers.
The return value 0 indicates the success on the request to clock driver,
and negative values indicate the failure reasons.

Signed-off-by: Terry Bai <terry.z.bai@gmail.com>

terryzbai committed:
clk: combine regmap_* and regmap_mux_* operations

Abstract a set of regmap_* operations by combining mux-specific ones and ordinary ones

Signed-off-by: Terry Bai <terry.z.bai@gmail.com>

terryzbai committed:
clk: add get/set parent interfaces

Add clk_get_parent() and clk_set_parent() interfaces for the clients.
Fixed several logic issues in related implementations. More clock
component definition checkings and fixup.

Signed-off-by: Terry Bai <terry.z.bai@gmail.com>

terryzbai committed:
clk: set_rate for maaxboard clocks

This commit adds set_rate() implementations for maaxboard. Now clocks can
be set to target rate at initialisation time and configured dynamically.

Signed-off-by: Terry Bai <terry.z.bai@gmail.com>

terryzbai committed:
clk: reorganise files

This commit reorgnises source files to allow more code to be shared
for multiple boards.

Refactor makefiles:
- move objectives to clk_driver directory
- use built-in CFLAGS in rules

Temporary fixup for a zig build issue: create_clk_config.py cannot find
the lazypath `zig-out/bin` when executing. The current solution is to
create an install directory in advance, but it would be a bit of trouble
to find the generated header file if someone want to read or modify it.

Moved clock specifier binding files to `include/sddf/clk`, so the clients
will be able to configure clocks with macro constants

Signed-off-by: Terry Bai <terry.z.bai@gmail.com>

omeh-a commented:
Please consider rebasing to squash down some of these commits, it's a bit tricky to handle in git when this feature is broken into 11

omeh-a commented on examples/clk/build.zig:
```diff
@@ -0,0 +1,147 @@
+
+    const dtb_arg = b.option([]const u8, "dtb", "Path to DTB file");
+    if (dtb_arg == null) {
+        std.log.err("Missing -Dsdk=/path/to/dtb argument being passed\n", .{});
```
Typo here - this should be `Missing -Ddtb` not `Missing -Dsdk`. This stuff should also be in a README.md for this example but there isn't one

omeh-a committed:
Update clock driver to current sddf build style, added meta.py

Signed-off-by: Lesley Rossouw <lesley.rossouw@unsw.edu.au>

omeh-a committed:
Replace setvar constants, move all memory region setup to meta.py

Signed-off-by: Lesley Rossouw <lesley.rossouw@unsw.edu.au>

terryzbai committed:
examples/clk: remove timer for client

Signed-off-by: Terry Bai <tianyi.bai@unsw.edu.au>

terryzbai committed:
examples/clk: remove dts dir

Signed-off-by: Terry Bai <tianyi.bai@unsw.edu.au>

terryzbai committed:
drivers/clk: add sddf_clk_set_cpu_freq() interface

Signed-off-by: Terry Bai <tianyi.bai@unsw.edu.au>

terryzbai committed:
drivers/clk: add ARM_CLK and remove sddf_clk_set_cpu_freq()

Signed-off-by: Terry Bai <tianyi.bai@unsw.edu.au>

midnightveil commented on drivers/clk/imx/include/clk-imx.h:
```diff
@@ -0,0 +1,397 @@
+    },                                                      \
+    .hw.init = &(struct clk_init_data) {                    \
+        .name = #_name,                                     \
+        .ops = &clk_divider_ops,                            \
```
I'm copying @wom-bat comment here btw from his original review, because it was never addressed.

> These are too close to the Linux drivers. We don't need to provide a run-time interface for different clocks, as any sDDF system runs only on a single SoC --- so get rid of all the `..._ops` structures and call the functions directly.

terryzbai committed:
drivers/clk: fix compilation complaints

Signed-off-by: Terry Bai <tianyi.bai@unsw.edu.au>

terryzbai committed:
drivers/clk: fix returned_freq of clk_set_rate()

Signed-off-by: Terry Bai <tianyi.bai@unsw.edu.au>

terryzbai committed:
drivers/clk: fix parent clock of ARM_CLK for imx8mq

Signed-off-by: Terry Bai <tianyi.bai@unsw.edu.au>

midnightveil commented on examples/clk/meta.py:
```diff
@@ -0,0 +1,121 @@
+)
+from board import BOARDS
+
+assert version("sdfgen").split(".")[1] == "28", "Unexpected sdfgen version"
```
remove https://github.com/au-ts/sddf/pull/709
